### ISSUE TITLE
Version 4.1.0 release broke authentication against B2C

### ISSUE THREAD

**Issue by rudfoss** [AzureAD/passport-azure-ad#425]
The 4.1.0 release broke authentication against B2C with policy. The problem specifically is this line:

```javascript
// lib/oidcstrategy.js:861
return u.pathname && _.includes(u.pathname, policy.toLowerCase());
```

This line is part of a function called `policyChecker` that attempts to determine whether the policy is part of the endpoint url. The problem is that the parsed url, the `u` variable pathname property does not include the query string at all.

If we compare this with the old version of the same file:

```
return u.query && u.query.p && (policy.toLowerCase() === u.query.p.toLowerCase());
```

We can see that the comparison is done quite differently. Here there is an actual check against the query parameters of the url and not just a general includes lookup on the string. As far as I can see the new implementation will never work correctly.

(edit for clarity)

**Comment by rudfoss** [comment]
I've created a pull request that resolves this issue here:
https://github.com/AzureAD/passport-azure-ad/pull/426

Though I've not had a chance to thoroughly test it on older versions of node (running 10.13.0)

**Comment by flyboarder** [comment]
Even with the patch, I am unable to authenticate due to `In collectInfoFromReq: policy is missing`.

It seems the policyName is never used, so how do we set the policy?

**Comment by rudfoss** [comment]
> Even with the patch, I am unable to authenticate due to `In collectInfoFromReq: policy is missing`.
> 
> It seems the policyName is never used, so how do we set the policy?

Did you set the query parameter p=[policyName] in the url?

**Comment by flyboarder** [comment]
Yes, I was eventually able to get it working in my case

**Comment by ptc-bsorek** [comment]
Hi, when will this fix be merged?

**Comment by kerosan** [comment]
any updates?

**Comment by tvvignesh** [comment]
I am facing the same issue as well documented by people in https://github.com/AzureAD/passport-azure-ad/issues/418 and https://github.com/AzureAD/passport-azure-ad/issues/425

I manually set the policy by setting `req.query.p='Mypolicyname';` before `passport.authenticate` but then next, I started getting the error `Cannot get AAD Federation metadata` I checked the identityMetadata url many times and it is right but figured that it was because of the line I added above due to which the library was calling `https://login.microsoftonline.com/mytenant.onmicrosoft.com/v2.0/.well-known/openid-configuration?x-client-SKU=passport-azure-ad&x-client-Ver=4.0.0&p=Mypolicyname` instead of `https://login.microsoftonline.com/mytenant.onmicrosoft.com/v2.0/.well-known/openid-configuration?x-client-SKU=passport-azure-ad&x-client-Ver=4.0.0` and the adding the parameter gives 404 not found.

Any idea how to solve both these issues? Struggling on this for a long time.

**Comment by tvvignesh** [comment]
@DarylThayil @lovemaths @brandwe Your intervention would be great. Thanks.

**Comment by DarylThayil** [comment]
Approved and merged

**Comment by DarylThayil** [comment]
released in 4.2.1
